### ISSUE TITLE
LDAP auth issue

### ISSUE THREAD

**Issue by kaos** [rancher/rancher#2191]
ref #2051 

This is strange. When we run a local instance on a couple of virtualbox hosted machines (using docker machine/boot2docker), it works just fine. Both when we enable LDAP auth before registering any hosts and after a first host has been registered.

But now, when we run on bare metal, using an external Maria DB and enable LDAP auth, we got issues. So the only differences I can see are those: Bare metal vs Virtualbox hosted machine, and using the mysql db shipped with rancher vs an external mariadb.

LDAP enable goes fine. The account that set it up can login just fine. But another user account trying to login gets a 500, with this message in the server logs:

```
2015-09-30 14:18:09,622 ERROR [:] [] [] [] [tp1167130569-16] [i.g.i.g.r.handler.ExceptionHandler  ] Exception in API for request [io.github.ibuildthecloud.gdapi.request.ApiRequest@49310717] java.lang.NullPointerException: null
	at io.cattle.platform.iaas.api.auth.integration.ldap.LdapIdentitySearchProvider.attributesToIdentity(LdapIdentitySearchProvider.java:378) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.iaas.api.auth.integration.ldap.LdapIdentitySearchProvider.getObject(LdapIdentitySearchProvider.java:302) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.iaas.api.auth.integration.ldap.LdapIdentitySearchProvider.getIdentities(LdapIdentitySearchProvider.java:207) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.iaas.api.auth.integration.ldap.LdapIdentitySearchProvider.getIdentities(LdapIdentitySearchProvider.java:233) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.iaas.api.auth.integration.ldap.LdapTokenCreator.getLdapToken(LdapTokenCreator.java:48) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.iaas.api.auth.integration.ldap.LdapTokenCreator.getToken(LdapTokenCreator.java:63) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.iaas.api.auth.identity.TokenResourceManager.createToken(TokenResourceManager.java:52) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.iaas.api.auth.identity.TokenResourceManager.createInternal(TokenResourceManager.java:41) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
	at io.github.ibuildthecloud.gdapi.request.resource.impl.AbstractBaseResourceManager.create(AbstractBaseResourceManager.java:115) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
	at io.github.ibuildthecloud.gdapi.request.handler.ResourceManagerRequestHandler.generate(ResourceManagerRequestHandler.java:39) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
	at io.github.ibuildthecloud.gdapi.request.handler.AbstractResponseGenerator.handle(AbstractResponseGenerator.java:14) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
	at io.github.ibuildthecloud.gdapi.request.handler.write.DefaultReadWriteApiDelegate.handle(DefaultReadWriteApiDelegate.java:27) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
	at io.github.ibuildthecloud.gdapi.request.handler.write.DefaultReadWriteApiDelegate.write(DefaultReadWriteApiDelegate.java:22) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_79]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_79]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_79]
	at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_79]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317) ~[spring-aop-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183) ~[spring-aop-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150) ~[spring-aop-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:96) ~[spring-tx-3.2.4.RELEASE.jar:3.2.4.RELEASE]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:260) ~[spring-tx-3.2.4.RELEASE.jar:3.2.4.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:94) ~[spring-tx-3.2.4.RELEASE.jar:3.2.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[spring-aop-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:91) ~[spring-aop-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[spring-aop-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204) ~[spring-aop-3.2.5.RELEASE.jar:3.2.5.RELEASE]
	at com.sun.proxy.$Proxy42.write(Unknown Source) ~[na:na]
	at io.github.ibuildthecloud.gdapi.request.handler.write.ReadWriteApiHandler.handle(ReadWriteApiHandler.java:19) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
	at io.github.ibuildthecloud.gdapi.servlet.ApiRequestFilterDelegate.doFilter(ApiRequestFilterDelegate.java:86) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.api.servlet.ApiRequestFilter$1.runInContext(ApiRequestFilter.java:79) [cattle-framework-api-0.5.0-SNAPSHOT.jar:na]
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49) [cattle-framework-managed-context-0.5.0-SNAPSHOT.jar:na]
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55) [cattle-framework-managed-context-0.5.0-SNAPSHOT.jar:na]
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:108) [cattle-framework-managed-context-0.5.0-SNAPSHOT.jar:na]
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52) [cattle-framework-managed-context-0.5.0-SNAPSHOT.jar:na]
	at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46) [cattle-framework-managed-context-0.5.0-SNAPSHOT.jar:na]
	at io.cattle.platform.api.servlet.ApiRequestFilter.doFilter(ApiRequestFilter.java:72) [cattle-framework-api-0.5.0-SNAPSHOT.jar:na]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) [jetty-servlet-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.servlets.UserAgentFilter.doFilter(UserAgentFilter.java:83) [jetty-servlets-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.servlets.GzipFilter.doFilter(GzipFilter.java:300) [jetty-servlets-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) [jetty-servlet-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585) [jetty-servlet-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577) [jetty-security-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515) [jetty-servlet-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.Server.handle(Server.java:499) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257) [jetty-server-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540) [jetty-io-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635) [jetty-util-9.2.11.v20150529.jar:9.2.11.v20150529]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555) [jetty-util-9.2.11.v20150529.jar:9.2.11.v20150529]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_79]
```

Both systems tested on a fresh install of v0.39.0.

**Comment by deniseschannon** [comment]
I haven't quite reproduced. 

Steps that I did:
Launch Rancher Server v0.39.0 linked to an external mariaDB container. 
Enable LDAP (non TLS) with userA.
Add userB to Default environment. 
Log in as user B to environment.

I had no issues logging in. I also had no hosts. 

Were you using TLS for LDAP?

**Comment by deniseschannon** [comment]
Also, did you add a group instead of a specific user? I believe there is an issue with it and this could be the issue to track it. :)

**Comment by kaos** [comment]
Ah, initially I made attempts both with groups and users. But in my last attempts with clean installs to try this out, I did neither. So, userA enables LDAP auth (non TLS), and userB tries to login.

I would expect userB to be able to login even tough not being a member of any environments. Failing that (which would be fine) I'd at least expect a not authorized message rather than the internal server error.

Will try again with adding userB to the environment.

Note however, that userB was able to login in our testing when running locally using the bundled mysql db, where as with external db, he was not.

**Comment by kaos** [comment]
Another point that may be relevant. In our external db case, it talks to a cluster. So if there is an update to the db, and a different connection is used to query data that relies on a previous update, it may be that it has not been synced yet.

Is it possible that such a race condition exists here?

**Comment by vincent99** [comment]
It is probably a race condition yes, but more like one that happens even with a single DB server but is only possible/easier to trigger with some latency between it and the server.  I "fixed" a similar issue in local auth last week with `setTimeout` between enabling and testing auth.  The person who worked on this will be back next week and looking into it.

When you say "cluster" what exactly do you mean?  Multi-master  synchronous replication (I.e. Percona XtraDB Cluster)?b what you described is asynchronous eventually consistent-style explication...  If it's really that I'm frankly surprised anything works at all, as we rather like the "C" and "D" in ACID :smile:.

**Comment by kaos** [comment]
I'll pass this question to one of our DB guys, who asked me questions regarding this, so I got the impression of my previous concerns. I do agree with that the ACID properties ought to take care of this..

**Comment by ulffernholm** [comment]
Hi!
we have MariaDB+GalareCluster and it is not exactly ACID behavior but "close", as the setup is now

**Comment by kaos** [comment]
@vincent99 I'm running tests now to bypass the cluster bit so we always talk to the same node to get ACID properties for rancher. If this turns out to solve the issue, we have plans for a more permanent solution.

**Comment by kaos** [comment]
It made no difference for us, so I'm under the impression that it's not related to the ACIDness of our setup. Will take a closer look on Monday at the actual data in the auth tables and compare it to what I have locally that works to see if I can spot any discrepancies.

**Comment by hibooboo2** [comment]
@kaos I was able to reproduce your stacktrace by making the name of the `Ldap Group Name Field` something that is an attribute that doesn't exist. rancher/cattle#936 this will prevent this stacktrace from occurring.  when it hits a release I will post the release so that you can retest this situation, as I was unable to try it using what I think is causing you problem which is the group having no name or name attribute on it; but the other groups having it which would be why one user can login and the other cannot.

**Comment by kaos** [comment]
@hibooboo2 Aaww.. well of course. Thank you! <3 <3 <3 I've yet to confirm this, but it's so obvious now that you've pointed it out for me. I was so stuck in the notion of that "it worked with the internal db, but not our external" that I completely overlooked the possibility that I didn't have the _exact_ same config in both cases apart from the db stuff.. even looking at the failing line from the backtrace now makes me wonder how I could've missed it. I've done these kinds of mistakes before, wonder how long before I actually learn from them :p

**Comment by kaos** [comment]
Confirmed. I changed our group name field to the default value of `name`, and now it works. As it turns out, not all of our groups had a `displayName` field... so, rancher/cattle#936 will resolve this for us. 

However, I wonder if it wouldn't be better to fallback to `name` rather than `dn` when the requested field is missing (perhaps with `dn` as a secondary fallback, in case `name` is missing too..).. ? @hibooboo2
